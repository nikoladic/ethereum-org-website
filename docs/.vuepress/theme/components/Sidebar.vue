<template>
  <div class="sidebar pa-1 hidden sm-up-block" id="outline">
    <slot name="top" />
    <ul class="sidebar-links no-bullets ma-0 pa-0 l8" v-if="items.length">
      <li v-for="(item, i) in items" :key="i">
        <SidebarGroup
          v-if="item.type === 'group'"
          :item="item"
          :first="i === 0"
          :open="i === openGroupIndex"
          :collapsable="item.collapsable || item.collapsible"
          @toggle="toggleGroup(i)"
        />
        <SidebarLink v-else :item="item" />
      </li>
    </ul>
    <slot name="bottom" />
  </div>
</template>

<script>
import SidebarGroup from './SidebarGroup.vue'
import SidebarLink from './SidebarLink.vue'
import { isActive } from '../utils/util'

export default {
  components: { SidebarGroup, SidebarLink },

  props: ['items'],

  data() {
    return {
      openGroupIndex: 0
    }
  },

  created() {
    this.refreshIndex()
  },

  watch: {
    $route() {
      this.refreshIndex()
    }
  },

  methods: {
    refreshIndex() {
      const index = resolveOpenGroupIndex(this.$route, this.items)
      if (index > -1) {
        this.openGroupIndex = index
      }
    },

    toggleGroup(index) {
      this.openGroupIndex = index === this.openGroupIndex ? -1 : index
    },

    isActive(page) {
      return isActive(this.$route, page.path)
    }
  }
}

function resolveOpenGroupIndex(route, items) {
  for (let i = 0; i < items.length; i++) {
    const item = items[i]
    if (
      item.type === 'group' &&
      item.children.some(c => isActive(route, c.path))
    ) {
      return i
    }
  }
  return -1
}
</script>

<style lang="stylus">
@import '../styles/config.styl'

.sidebar
  position sticky
  top 72px
  bottom 0
  right 0
  width $sidebarWidth
  height calc(100vh - 80px)
  overflow-y auto
  transition all 0.2s ease-in-out
  transition transform .2s ease

.sidebar-links
  border-left 1px solid $colorWhite700
.dark-mode
  .sidebar-links
    border-left 1px solid $colorBlack200
</style>
