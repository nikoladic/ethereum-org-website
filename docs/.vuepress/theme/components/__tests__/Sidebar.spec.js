import { shallowMount } from '@vue/test-utils'
import Sidebar from "@/theme/components/Sidebar"

describe('Sidebar', () => {
  test('render sidebar groups', () => {
    const wrapper = shallowMount(Sidebar, {
      propsData: {
        items: [
          {
            title: "group1",
            children: ["a", "b"]
          },
          {
            title: "group2",
            children: []
          },
        ]
      },
      mocks: {
        $route: {}
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  test('render no sidebar groups', () => {
    const wrapper = shallowMount(Sidebar, {
      propsData: {
        items: []
      },
      mocks: {
        $route: {}
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})

