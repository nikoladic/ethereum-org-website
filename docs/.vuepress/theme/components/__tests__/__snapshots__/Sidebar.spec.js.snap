// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar render empty sidebar 1`] = `
<div
  class="sidebar"
>
  <navlinks-stub
    issidebar="true"
  />
    
  <!---->
   
</div>
`;

exports[`Sidebar render sidebar groups 1`] = `
<div
  class="sidebar"
>
  <navlinks-stub
    issidebar="true"
  />
    
  <ul
    class="sidebar-links"
  >
    <li>
      <sidebargroup-stub
        first="true"
        item="[object Object]"
        open="true"
      />
    </li>
    <li>
      <sidebarlink-stub
        item="[object Object]"
      />
    </li>
  </ul>
   
</div>
`;

exports[`Sidebar render sidebar items 1`] = `
<div
  class="sidebar"
>
  <navlinks-stub
    issidebar="true"
  />
    
  <ul
    class="sidebar-links"
  >
    <li>
      <sidebarlink-stub
        item="[object Object]"
      />
    </li>
    <li>
      <sidebarlink-stub
        item="[object Object]"
      />
    </li>
  </ul>
   
</div>
`;
